Source/WebCore/Modules/webcodecs/WebCodecsAudioDecoder.cpp

```diff
@@ -0,0 +1,323 @@
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebCodecsAudioDecoder.h"
+
+#if ENABLE(WEB_CODECS)
+
+#include "DOMException.h"
+#include "Event.h"
+#include "EventNames.h"
+#include "JSDOMPromiseDeferred.h"
+#include "JSWebCodecsAudioDecoderSupport.h"
+#include "ScriptExecutionContext.h"
+#include "WebCodecsAudioData.h"
+#include "WebCodecsAudioDataOutputCallback.h"
+#include "WebCodecsEncodedAudioChunk.h"
+#include "WebCodecsErrorCallback.h"
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(WebCodecsAudioDecoder);
+
+Ref<WebCodecsAudioDecoder> WebCodecsAudioDecoder::create(ScriptExecutionContext& context, Init&& init)
+{
+    auto decoder = adoptRef(*new WebCodecsAudioDecoder(context, WTFMove(init)));
+    decoder->suspendIfNeeded();
+    return decoder;
+}
+
+WebCodecsAudioDecoder::WebCodecsAudioDecoder(ScriptExecutionContext& context, Init&& init)
+    : ActiveDOMObject(&context)
+    , m_output(init.output.releaseNonNull())
+    , m_error(init.error.releaseNonNull())
+{
+}
+
+WebCodecsAudioDecoder::~WebCodecsAudioDecoder()
+{
+}
+
+static bool isValidDecoderConfig(const WebCodecsAudioDecoderConfig& config, const Settings::Values&)
+{
+    // FIXME: Check codec more accurately.
+    if (!config.codec.startsWith("mp4a."_s) && config.codec != "mp3"_s && config.codec != "opus"_s && config.codec != "alaw"_s && config.codec != "ulaw"_s && config.codec != "flac"_s && config.codec != "vorbis"_s && !config.codec.startsWith("pcm-"_s))
+        return false;
+
+    return true;
+}
+
+static AudioDecoder::Config createAudioDecoderConfig(const WebCodecsAudioDecoderConfig& config)
+{
+    std::span<const uint8_t> description;
+    if (config.description) {
+        auto* data = std::visit([](auto& buffer) -> const uint8_t* {
+            return buffer ? static_cast<const uint8_t*>(buffer->data()) : nullptr;
+        }, *config.description);
+        auto length = std::visit([](auto& buffer) -> size_t {
+            return buffer ? buffer->byteLength() : 0;
+        }, *config.description);
+        if (length)
+            description = { data, length };
+    }
+
+    return {
+        description,
+        config.sampleRate.value_or(0),
+        config.numberOfChannels.value_or(0)
+    };
+}
+
+ExceptionOr<void> WebCodecsAudioDecoder::configure(ScriptExecutionContext& context, WebCodecsAudioDecoderConfig&& config)
+{
+    if (!isValidDecoderConfig(config, context.settingsValues()))
+        return Exception { TypeError, "Config is not valid"_s };
```

Comment:
This is fine as is, but the spec is moving now to close the decoder with NotSupportedError instead.
I'll do the change for video decoder/encoder this week.